warehouse/observations/models.py

```diff
@@ -63,7 +63,7 @@ class Observer(db.Model):
     parent: AssociationProxy = association_proxy("_association", "parent")
 
 
-class HasObserversMixin:
+class HasObservers:
```

Comment:
Sorry for a noob question - Would this model change require a migration? Does Pyramid have this concept or are there other things managing that?

Reply:
> Would this model change require a migration?

Good question! Some model changes, like adding/removing attributes or modifying the attributes, do require a migration.

This change doesn't require a migration because a) the classname of the mixin is only used in Python and doesn't extend into the database and b) the `@declared_attr` that I'm removing is not a "real" attribute in the sense that it has a column in the database, it's just injected in Python.

We can confirm this by attempting to autogenerate a migration:

```
$ docker-compose run web python -m warehouse db revision --autogenerate --message "Does this need a migration?"
Creating warehouse_web_run ... done
{"logger": "alembic.runtime.migration", "level": "INFO", "event": "Context impl PostgresqlImpl.", "thread": 140295611669376}
{"logger": "alembic.runtime.migration", "level": "INFO", "event": "Will assume transactional DDL.", "thread": 140295611669376}
{"logger": "alembic.ddl.postgresql", "level": "INFO", "event": "Detected sequence named 'file_registry_id_seq' as owned by integer column 'file_registry(id)', assuming SERIAL and omitting", "thread": 140295611669376}
{"logger": "alembic.ddl.postgresql", "level": "INFO", "event": "Detected sequence named 'journals_id_seq' as owned by integer column 'journals(id)', assuming SERIAL and omitting", "thread": 140295611669376}
{"logger": "alembic.ddl.postgresql", "level": "INFO", "event": "Detected sequence named 'accounts_email_id_seq' as owned by integer column 'user_emails(id)', assuming SERIAL and omitting", "thread": 140295611669376}
  Generating /opt/warehouse/src/warehouse/migrations/versions/bf8094851d63_does_this_need_a_migration.py ...  done
  Running post write hook 'black' ...
reformatted /opt/warehouse/src/warehouse/migrations/versions/bf8094851d63_does_this_need_a_migration.py

All done! ✨ 🍰 ✨
1 file reformatted.
  done
  Running post write hook 'isort' ...
Fixing /opt/warehouse/src/warehouse/migrations/versions/bf8094851d63_does_this_need_a_migration.py
  done
  ```
  
  And seeing if it has any contents:
  
  ```
  $ cat warehouse/migrations/versions/bf8094851d63_does_this_need_a_migration.py
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""
Does this need a migration?

Revision ID: bf8094851d63
Revises: 14ad61e054cf
Create Date: 2024-07-09 17:32:03.675839
"""

import sqlalchemy as sa

from alembic import op

revision = "bf8094851d63"
down_revision = "14ad61e054cf"

# Note: It is VERY important to ensure that a migration does not lock for a
#       long period of time and to ensure that each individual migration does
#       not break compatibility with the *previous* version of the code base.
#       This is because the migrations will be ran automatically as part of the
#       deployment process, but while the previous version of the code is still
#       up and running. Thus backwards incompatible changes must be broken up
#       over multiple migrations inside of multiple pull requests in order to
#       phase them in over multiple deploys.
#
#       By default, migrations cannot wait more than 4s on acquiring a lock
#       and each individual statement cannot take more than 5s. This helps
#       prevent situations where a slow migration takes the entire site down.
#
#       If you need to increase this timeout for a migration, you can do so
#       by adding:
#
#           op.execute("SET statement_timeout = 5000")
#           op.execute("SET lock_timeout = 4000")
#
#       To whatever values are reasonable for this migration as part of your
#       migration.


def upgrade():
    # ### commands auto generated by Alembic - please adjust! ###
    pass
    # ### end Alembic commands ###


def downgrade():
    # ### commands auto generated by Alembic - please adjust! ###
    pass
 ```
  
 Since both `upgrade` and `downgrade` are empty, no migration is necessary.

> Does Pyramid have this concept or are there other things managing that?

I think you're talking about managing database migrations, correct? We use `alembic` for that, more details here: https://warehouse.pypa.io/development/database-migrations.html

Reply:
thanks for imparting your knowledge :P